api/desecapi/serializers.py

```diff
@@ -600,10 +603,89 @@ def validate_name(self, value):
             raise serializers.ValidationError(self.default_error_messages['name_unavailable'], code='name_unavailable')
         return value
 
+    def parse_zonefile(self, domain_name: str, zonefile: str):
+        try:
+            self.import_zone = dns.zone.from_text(
+                zonefile,
+                origin=dns.name.from_text(domain_name),
+                allow_include=False,
+                check_origin=False
+            )
+        except dns.zonefile.CNAMEAndOtherData:
+            raise serializers.ValidationError(
+                {'zonefile': ['No other records with the same name are allowed alongside a CNAME record.']})
+        except ValueError as e:
+            if 'has non-origin SOA' in str(e):
+                raise serializers.ValidationError(
+                    {'zonefile': [f'Zonefile includes an SOA record for a name different from {domain_name}.']})
+            raise e
+        except dns.exception.SyntaxError as e:
+            try:
+                line = str(e).split(':')[1]
+                raise serializers.ValidationError({'zonefile': [f'Zonefile contains syntax error in line {line}.']})
+            except IndexError:
+                raise serializers.ValidationError({'zonefile': [f'Could not parse zonefile: {str(e)}']})
+
+    def validate(self, attrs):
+        if attrs.get('zonefile') is not None:
+            self.parse_zonefile(attrs.get('name'), attrs.pop('zonefile'))
+        return super().validate(attrs)
+
+    def to_internal_value(self, data):
+        return super().to_internal_value(data)
+
     def create(self, validated_data):
+        # save domain
         if 'minimum_ttl' not in validated_data and models.Domain(name=validated_data['name']).is_locally_registrable:
             validated_data.update(minimum_ttl=60)
-        return super().create(validated_data)
+        domain: models.Domain = super().create(validated_data)
+
+        # save RRsets if zonefile was given
+        if self.import_zone:
```

Comment:
It is initialized to `None` and only set if `parse_zonefile` is called and does not error. `parse_zonefile` is only called if `attrs.get('zonefile') is not None`.

Reply:
`if getattr(self.import_zone, 'nodes', None)`?

Reply:
... or:
```python
nodes = getattr(self.import_zone, 'nodes', None)  # or try self.import_zone.nodes, except assign None
if nodes:
   # and then use `nodes` instead of `self.import_zone.nodes`
```

Reply:
done